### PR TITLE
Add license parameter

### LINKED ISSUE #4: Add license parameter

### PR DESCRIPTION
**Pull request by paw-lu**
This PR adds a `license` parameter. Currently there are three options: MIT, Apache-2.0, and GPL-3.0. All references to MIT and MIT License throughout the template are replaced with `cookiecutter.license` and `cookiecutter.full_license_name`, respectively.

Additionally two new parameters are added to the template—`year` and `full_license_name`. I had originally meant for these two variables to be [private rendered variables that begin with `__`](https://cookiecutter.readthedocs.io/en/latest/advanced/private_variables.html#private-variables) so the user does not interact with them, but it seems that this feature is not yet released.

The logic for `full_license_name` is honestly a bit ugly.

The `TimeExtension` is [one of the template extensions included with Cookiecutter](https://cookiecutter.readthedocs.io/en/latest/advanced/template_extensions.html), so it will not require the user to install any additional dependencies.

Closes #4

### CODE REVIEW

**Comment by paw-lu** [comment]
[GitHub previews of the added license files are hosted as a gist.](https://gist.github.com/paw-lu/840bc43141584f5d62219548d4cfca25)

The Apache-2.0 text was written by me and the [GPL-3.0 was taken directly from gnu.org.](https://www.gnu.org/licenses/gpl-3.0.rst)

File: cookiecutter.json [hunk 1]
```diff
@@ -5,5 +5,9 @@
   "author": "Claudio Jolowicz",
   "email": "mail@claudiojolowicz.com",
   "github_user": "cjolowicz",
-  "version": "0.0.0"
+  "version": "0.0.0",
+  "license": ["MIT", "Apache-2.0", "GPL-3.0"],
+  "year": "{% now 'utc', '%Y' %}",
+  "full_license_name": "{{ {'MIT': 'MIT License', 'Apache-2.0': 'Apache Licence 2.0', 'GPL-3.0': 'Gnu General Public License v3.0'}.get(cookiecutter.license, '') }}",
+  "_extensions": ["jinja2_time.TimeExtension"]
```
Comment (paw-lu) [inline cookiecutter.json:12]: This is the line Prettier has issues with, but it seems to match the desired output.

**Comment by paw-lu** [comment]
Nox's pre-commit run is causing is causing the checks to fail. The issue seems to be with prettier, but only on Python 3.9. Prettier had passed locally.

Curiously, the output diff doesn't make sense to me 😅

```txt
nox > pre-commit run --all-files --show-diff-on-failure
black....................................................................Passed
Check for added large files..............................................Passed
Check Toml...............................................................Passed
Check Yaml...............................................................Passed
Fix End of Files.........................................................Passed
flake8...................................................................Passed
Reorder python imports...................................................Passed
Trim Trailing Whitespace.................................................Passed
prettier.................................................................Failed
- hook id: prettier
- files were modified by this hook

.cookiecutter.json

pre-commit hook(s) made changes.
If you are seeing this message in CI, reproduce locally with: `pre-commit run --all-files`.
To run `pre-commit` as part of git workflow, use `pre-commit install`.
All changes made by hooks:
diff --git a/.cookiecutter.json b/.cookiecutter.json
index f0d8565..1379d39 100644
--- a/.cookiecutter.json
+++ b/.cookiecutter.json
@@ -1,7 +1,5 @@
 {
-  "_extensions": [
-    "jinja2_time.TimeExtension"
-  ],
+  "_extensions": ["jinja2_time.TimeExtension"],
   "_template": "cookiecutter-hypermodern-python",
   "author": "Claudio Jolowicz",
   "email": "mail@claudiojolowicz.com",
```

[Because if we look at that line](https://github.com/cjolowicz/cookiecutter-hypermodern-python/pull/701/files#r543038057) , it does not match the removed portion of the diff, but does match the reported "change".

Will investigate later!

**Comment by paw-lu** [comment]
Realized I was mistaken!

Prettier is complaining about the _generated_ `.cookiecutter.json` in the project itself, as opposed to the template's `cookiecutter.json`.

File: cookiecutter.json (hunk 1 above)
Comment (paw-lu) [reply cookiecutter.json:12]: This is incorrect:

> [Prettier is complaining about the generated .cookiecutter.json in the project itself, as opposed to the template's cookiecutter.json.](https://github.com/cjolowicz/cookiecutter-hypermodern-python/pull/701#issuecomment-745587106)

**Comment by paw-lu** [comment]
[While the documentation suggests otherwise](https://cookiecutter.readthedocs.io/en/latest/advanced/template_extensions.html), it seems you don't need to specify the included extensions.

In fact I noticed we were using the `cookiecutter.extensions.JsonifyExtension` extension in `.cookiecutter.json` without specifying it in the `_extensions` variable.

I'm not sure how fragile this is as it's undocumented (will you need to specify extensions in the future?).

The latest commit removes the `_extensions` variable—causing all tests to pass, though this issue might resurface.

Another option would be to add a `.prettierignore`.

```txt
.cookiecutter.json
```

---

Am not sure why the 3.6–3.8 tests passed though 🤔

**Review by cjolowicz — CHANGES_REQUESTED** [review]
Thank you for your contribution, this looks nice!

I think we can get rid of the `year` variable, and inline the tag where needed. See how [pydanny/cookiecutter-django] do this, for example.

[pydanny/cookiecutter-django]: https://github.com/pydanny/cookiecutter-django/blob/55003e21a62c761910f64a4899842a986d5fdfba/%7B%7Bcookiecutter.project_slug%7D%7D/LICENSE 

In fact, I would somewhat prefer to have only `license` as a Cookiecutter variable, and replace occurrences of `{{ cookiecutter.full_license_name }}` with `{{ cookiecutter.license.replace("-", " ") }} license`, even if it does not look quite as nice in the rendered output.

Your solution avoids having a hook moving or removing license files, and I like that. Recently I noticed there is a third way to deal with this problem, and I thought I'd mention it here: You could have separate license files with conditionals in their names, like this for the MIT license: `{% if cookiecutter.license == 'MIT' %}LICENSE.rst{% endif %}`. Files with an empty name will be ignored by Cookiecutter. Not sure though if that solution is actually better than what we have here, and quite happy to leave this as it is.

**Comment by paw-lu** [comment]
Ended up aggreeing with all of your suggestions, and implimented them.

I had originally written up this PR under the assumption that [cookiecutter private variables](https://cookiecutter.readthedocs.io/en/latest/advanced/private_variables.html) were available—so variables such as `year` and `full_license_name` were not exposed to the user—but it seems this is for the yet to be released `2.0.0` veersion of cookiecutter. When that feture is launched it might be worth implimenting them again as private variables.

> You could have separate license files with conditionals in their names, like this for the MIT license: `{% if cookiecutter.license == 'MIT' %}LICENSE.rst{% endif %}`.

Did not know this! Thanks for letting me know. I went ahead and implimented this. I think both have their (big) cons. Everything in one file—like I originally had—makes it difficult to parse the Jinja logic. On the other end, having multiple files can lead to a cluttered and confusing project root. For now we only have three `LICENSE.rst` variants—so I think that's okay. I have no problem reverting to one file if things get too ugly.

**Comment by paw-lu** [comment]
I see that the Windows test is failing now [due to an invalid path](https://github.com/cjolowicz/cookiecutter-hypermodern-python/pull/701/checks?check_run_id=1932896380#step:2:189) because of the Jinja-y file naming 😅.

**Comment by paw-lu** [comment]
`"` is an invalid character for file names in Windows. Changing to `'`.

File: {{cookiecutter.project_name}}/CONTRIBUTING.rst [hunk 2]
```diff
@@ -2,7 +2,7 @@ Contributor Guide
 =================
 
 Thank you for your interest in improving this project.
-This project is open-source under the `MIT license`_ and
+This project is open-source under the `{{cookiecutter.license.replace("-", " ")}}`_ and
```
Comment (cjolowicz) [inline {{cookiecutter.project_name}}/CONTRIBUTING.rst:5]:
```suggestion
This project is open-source under the `{{cookiecutter.license.replace("-", " ") license}} <{{cookiecutter.license}}>`_ and
```

File: {{cookiecutter.project_name}}/CONTRIBUTING.rst [hunk 3]
```diff
@@ -12,7 +12,7 @@ Here is a list of important resources for contributors:
 - `Issue Tracker`_
 - `Code of Conduct`_
 
-.. _MIT license: https://opensource.org/licenses/MIT
+.. _{{cookiecutter.license.replace("-", " ")}}: https://opensource.org/licenses/{{cookiecutter.license}}
```
Comment (cjolowicz) [inline {{cookiecutter.project_name}}/CONTRIBUTING.rst:15]:
```suggestion
.. _{{cookiecutter.license}}: https://opensource.org/licenses/{{cookiecutter.license}}
```

File: {{cookiecutter.project_name}}/README.rst [hunk 4]
```diff
@@ -74,7 +74,7 @@ To learn more, see the `Contributor Guide`_.
 License
 -------
 
-Distributed under the terms of the MIT_ license,
+Distributed under the terms of the `{{cookiecutter.license.replace("-", " ")}}`_,
```
Comment (cjolowicz) [inline {{cookiecutter.project_name}}/README.rst:77]:
```suggestion
Distributed under the terms of the `{{cookiecutter.license.replace("-", " ")}} license <{{cookiecutter.license}}>`_,
```

File: {{cookiecutter.project_name}}/README.rst [hunk 5]
```diff
@@ -93,7 +93,7 @@ This project was generated from `@cjolowicz`_'s `Hypermodern Python Cookiecutter
 
 .. _@cjolowicz: https://github.com/cjolowicz
 .. _Cookiecutter: https://github.com/audreyr/cookiecutter
-.. _MIT: http://opensource.org/licenses/MIT
+.. _{{cookiecutter.license.replace("-", " ")}}: https://opensource.org/licenses/{{cookiecutter.license}}
```
Comment (cjolowicz) [inline {{cookiecutter.project_name}}/README.rst:96]:
```suggestion
.. _{{cookiecutter.license}}: https://opensource.org/licenses/{{cookiecutter.license}}
```

**Review by cjolowicz — COMMENTED** [review]
Thanks for the update! I left some small suggestions inline, and some more thoughts below.

This PR will require some changes in the User Guide, specifically under *Creating a project* (the list of project variables), and *Documentation* (the description of `LICENSE.rst`). Possibly also the section *Files and directories*, for the documentation files, depending on how you handle the GPL (see below).

One thing I noticed is that for GPL-3.0 and Apache-2.0, the license text does not contain an actual copyright notice, only advice for how to add it to the project. I took a quick look at how [pydanny/cookiecutter-django] and [ionelmc/cookiecutter-pylibrary] handle this. The Django template places the copyright notice in `LICENSE` and includes the full text as the traditional `COPYING` file; but it only does this for the GPL. Ionel's template precedes the full license text by the copyright notice and a horizontal separator. Personally, I like Ionel's approach because it's pragmatic without losing clarity. What do you think?

**Comment by paw-lu** [comment]
Thanks for the suggestions! I got a little sloppy there with the replacement 😅. Thanks for catching that awkard wording. I don't think those result in valid reStructuredText links though—so I modified them a bit.

> This PR will require some changes in the User Guide, specifically under Creating a project (the list of project variables), and Documentation (the description of LICENSE.rst).

Good catch!

> One thing I noticed is that for GPL-3.0 and Apache-2.0, the license text does not contain an actual copyright notice, only advice for how to add it to the project.

Good eye. I actually based my implimentation based on what GitHub does by default if you select GPL-3.0 or Apache-2.0 as the license for the repo, so it's interesting that GitHub itself doesn't actually apply the notice. I placed the notice right before the full text on both files.

**Comment by cjolowicz** [comment]
LGTM :100: Can you rebase on `main`? That should take care of the CI failure.

**Comment by paw-lu** [comment]
Rebased!

**Review by cjolowicz — APPROVED** [review]
Thank you!

**Pull request merged**